### PR TITLE
Modified Linear Regression to work on OLS, fixes #8847

### LINKED ISSUE #8847: `machine_learning/linear_regression.py` doesn't give optimal coefficients
### Feature description

Related to issue #8844

TL;DR: the current implementation doesn't give optimal solutions, the current implementation calculates SSE wrong, and we should add an implementation of a numerical methods algorithm that actually gives optimal solutions

In `machine_learning/linear_regression.py`, add the following code at the bottom of the `main()` function:
```py
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt

data = np.asarray(data.astype(float))
X = data[:, 0].reshape(-1, 1)
y = data[:, 1]

reg = LinearRegression().fit(X, y)
print(f"Sklearn coefficients: {reg.intercept_}, {reg.coef_}")

sse = np.sum(np.square(reg.predict(X) - y.T))
print(f"{sse = }")
print(f"mse = {sse / len(y)}")
print(f"half mse = {sse / (2 * len(y))}")

plt.scatter(X, y, color="lightgray")
plt.axline(xy1=(0, theta[0, 0]), slope=theta[0, 1], color="red", label="Gradient descent")
plt.axline(xy1=(0, reg.intercept_), slope=reg.coef_, color="blue", label="Sklearn")
plt.legend()
plt.show()
```
This code performs ordinary least squares (OLS) linear regression using `sklearn` as a point of reference to compare the current implementation against. It then calculates the sum of squared errors (SSE), mean squared error (MSE), and half of the MSE. To compare the outputs visually, the code uses `matplotlib` to plot the `sklearn` regression line and the regression line produced by the current implementation.

The code produces the following command line output:
```
...
At Iteration 100000 - Error is 128.03882
Resultant Feature vector: 
-9.34325
1.53067
Sklearn coefficients: -15.547901662158367, [1.6076036]
sse = 253795.17406773588
mse = 253.79517406773587
half mse = 126.89758703386794
```
As we can see, what the implementation calculates as the SSE (128.03882) is actually half of the MSE, meaning that the `sum_of_square_error` function is incorrect and needs to be fixed. Why the implementation is using _half_ of the MSE, I have no clue.

Furthermore, we can see that both the regression coefficients and the errors are slightly off. This is because the current implementation works via gradient descent, meaning that it can only _approximate_ the OLS regression coefficients. Meanwhile, libraries like `numpy` and `sklearn` calculate the mathematically optimal coefficients using [numerical methods](https://en.wikipedia.org/wiki/Numerical_methods_for_linear_least_squares).
linear_regression_plot
Although using gradient descent to perform linear regression does work, it's still suboptimal and (AFAIK) it's not how linear regression is actually performed in practice. We can still include this implementation, but we should definitely also include an implementation of an optimal numerical method.

### PR DESCRIPTION
**Pull request by smruthi-sumanth**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Review by imSanko — APPROVED** [review]
Nice!

File: machine_learning/linear_regression.py [hunk 1]
```diff
@@ -31,85 +31,32 @@ def collect_dataset():
...
-    sum_elem = np.sum(np.square(prod))
-    error = sum_elem / (2 * len_data)
-    return error
-
-
-def run_linear_regression(data_x, data_y):
-    """Implement Linear regression over the dataset
-    :param data_x  : contains our dataset
-    :param data_y  : contains the output (result vector)
+def run_linear_regression_ols(data_x, data_y):
```
Comment (tianyizheng02) [inline machine_learning/linear_regression.py:34]:
```suggestion
def ols_linear_regression(data_x: np.ndarray, data_y: np.ndarray) -> np.ndarray:
```
1. Shortened the function name a bit (make sure you change the name elsewhere as well)
2. Added type hints

File: machine_learning/linear_regression.py [hunk 2]
```diff
@@ -31,85 +31,32 @@ def collect_dataset():
...
+    """Implement Linear regression using OLS over the dataset
```
Comment (tianyizheng02) [inline machine_learning/linear_regression.py:35]:
```suggestion
    """Implement OLS linear regression over a given dataset
```
Slight rewording for clarity

File: machine_learning/linear_regression.py [hunk 3]
```diff
@@ -31,85 +31,32 @@ def collect_dataset():
...
+    data_x = np.c_[np.ones(data_x.shape[0]), data_x].astype(float)
 
-    theta = np.zeros((1, no_features))
-
-    for i in range(iterations):
-        theta = run_steep_gradient_descent(data_x, data_y, len_data, alpha, theta)
-        error = sum_of_square_error(data_x, data_y, len_data, theta)
-        print(f"At Iteration {i + 1} - Error is {error:.5f}")
+    # Use NumPy's built-in function to solve the linear regression problem
+    theta = np.linalg.inv(data_x.T.dot(data_x)).dot(data_x.T).dot(data_y)
```
Comment (tianyizheng02) [inline machine_learning/linear_regression.py:44]:
```suggestion
    theta = np.linalg.inv(data_x.T @ data_x) @ data_x.T @ data_y
```
Instead of using `.dot()` for matrix multiplication, we can use numpy's `@` operator, which does the same thing and is more readable

File: machine_learning/linear_regression.py [hunk 4]
```diff
@@ -31,85 +31,32 @@ def collect_dataset():
...
-def run_linear_regression(data_x, data_y):
-    """Implement Linear regression over the dataset
-    :param data_x  : contains our dataset
-    :param data_y  : contains the output (result vector)
+def run_linear_regression_ols(data_x, data_y):
+    """Implement Linear regression using OLS over the dataset
+    :param data_x : contains our dataset
+    :param data_y : contains the output (result vector)
     :return        : feature for line of best fit (Feature vector)
     """
```
Comment (tianyizheng02) [inline machine_learning/linear_regression.py:39]: The OLS regression function needs doctests—make sure you verify the outputs of your tests with a calculator that can do linear regression (e.g., Wolfram Alpha)

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]
Good stuff! Apart from the changes that I've requested in my other comments, could you also rewrite the explanation at the top of the file to reflect your new implementation? In particular, make sure your new explanation does the following:
1. Briefly explain what linear regression is
2. Explain the OLS regression formula, $(\mathbf{X}^{\top} \mathbf{X})^{-1} \mathbf{X} \mathbf{y}$
3. Note possible issues with your implementation (i.e. inefficiency, numerical instability, etc.)
4. Cite a source that readers can refer to for further info (Wikipedia or similar is fine)

I wrote a similar explanation for weighted regression in [`machine_learning/local_weighted_learning/local_weighted_learning.py`](https://github.com/TheAlgorithms/Python/blob/edee8e644b09a21a1f70d3a59d57feed51c74004/machine_learning/local_weighted_learning/local_weighted_learning.py) a while back. If you'd like, you're welcome to use that explanation as a reference (though your explanation doesn't need to be nearly as long).

**Comment by cclauss** [comment]
Closing tests_are_failing PRs to prepare for Hacktoberfest 2024
